Making some things more explicit in the security r&d swe test (#58)
snyk/jobs: exercises/security-rnd/software-engineer/README.md

#### Things to consider

>   1. Look at the inner "dependencies" object for analysis of its dependencies and do so recursively.

Reviewer: I'm re-reading this compared to the original and can see what the original author meant. But also how it can be confusing.

Reviewer 2: After reading it 5 times, I can't understand the original meaning 😅

Reviewer: Basically - use npms API response to get the first level of dependencies. The implication was that you should then use that to get the next level recursively. It was just never explicitly said.

Reviewer 3: By explicitly saying recursive, i think we kind of missing the point of the test. Maybe we can go with something a bit more vague, like:
```suggestion
  1. Look at the inner "dependencies" object for analysis of all of its dependencies.
```

Reviewer: I think this is the part of the exercise we should be a bit more explicit about. It's confused literally every single applicant to the groot pipeline we had.

It seems to be particularly easy to misunderstand if English is not your first language

Reviewer 3: I definitely agree this is the most important part to improve and make a bit more explicit, but mentioning `recursive` makes this a no brainer to how to implement. I think we should try to emphasise the needs/expected outcome much clearer, but leave the implementation a bit vague to have some differences between candidates, specifically about how they interpreted it.

  2. There are currently over 800K packages on npmjs.com, and the number is growing all the time.
  3. The packages update from time to time, just as their dependencies.
  4. What makes a good web service? API, architecture, data storage, low latency, scalability, monitoring, you name it :)
  5. Consider the quality and structure of your codebase; is it maintainable?

>   6. Consider production readiness (to some extent) 

Reviewer: I'm not opposed to removing to `is it safe to deploy changes?`. But is there a strong reason for doing it?

Author: For me consider production readiness in the context of this home work is: make your application in a way that it can be packaged and delivered easily: docker, virtualenv, etc...

How would you evaluate that it is "safe to deploy changes" in the context of this home work?

Reviewer: Yeah I think thats a fair point - I can see how it can be confusing. I wonder what the original reason this was included was. Maybe @tomersnyk knows?

Reviewer 3: Happy to remove this part, but lets also remove the `to some extent` comment in that case:
```suggestion
  6. Consider production readiness
```

 

#### Implementation 

>   1. A working web application that, given a name of an npm published package, returns the set of the (recursive) dependencies for said package. Consider that different types of clients can make use of this endpoint.

Reviewer 3: ```suggestion
  1. A working web application that, given a name of an npm published package, returns all of its dependency tree.
```

Reviewer: I think"retrieve all the dependency tree" is clear enough 👍🏼

  2. The endpoint should return the recursive dependencies of a package as a dependency tree - should you get the whole dependency tree all at once or not? Why? Please justify this in your README

>   3. Account of asynchronous fetching of dependencies as you see fit.

Reviewer: I think we should remove this. As I explained in the google doc, I think this is just an artifact from copying the exercise over from the node/typescript version.

I have a feeling its confusing a lot of our candidates especially because we strongly suggest them to use python as the language for implementation.

Author: Fair, but when implementing the solution to get the whole dependency tree, this should be async; otherwise each request will only happen after all the other previous requests have finished; although this might be also something that can be improved in the second round of the interview?

Because now we're being explicit about the fact that we need the whole dependency tree in the response

Reviewer 2: Yes, please remove the reference to asynchronous. 

What matters, and could be a good discussion point for the interview, are performances. If the application works well it could be because caching / multi processing / async (or all of them :D )

Reviewer: > Fair, but when implementing the solution to get the whole dependency tree, this should be async; otherwise each request will only happen after all the other previous requests have finished;

True, but there could be other ways round that. In the end, serially requesting the data is a perfectly reasonable approach if you can justify it with the correct cache setup. 

My exercise worked this way for example. I justified it because I reckoned the cost for introducing async python (especially back when it was still very new) was not worth it when a cache layer which was warmed up with handle the vast majority of cases.

We should basically leave it up to the candidate to decide, justify and explain their reasoning without pushing them to a specific technology

Reviewer 3: ```suggestion
  3. Account the performance of fetching the dependencies as you see fit. 
```

  4. Cache relevant data so that repeated requests resolve with minimum latency

>   5. Add a few tests on the backend side

Reviewer: I'm not 100% sure about this.

I can see why its useful to point out - but at the same time I think it might be a useful indication of what the candidates prioritises

What do you think?

Reviewer 3: I think we shouldn't ask upfront to implement tests as we will be missing the point of understanding what the candidates finds valuable.

```suggestion
```

>   6. Add a README with instructions on how to install, run and test your code. Please, elaborate on your technical decisions you considered along the way.

Reviewer 3: Same here, we will miss the benefit of seeing the candidate decision making and prioritisation in a time box 
```suggestion
```

>   **We strongly suggest to implement this task in Python, as this is the main language used by our team, but we are happy for you to use another language if you feel it would allow you to better express your skillset. Please to make sure to give detailed instructions about how to use your code if you do decide to use another language**

Reviewer: nice :)

Reviewer 2: nice but I expect detailed instructions even for `python`, I don't want this point to be potentially understood as "well, they're python ninjas, they will figure it out" 😅

Reviewer: Yep, I think that is a fair point
